### ISSUE TITLE
optimizer and lr_scheduler starts with a learning rate different from the one set on command line

### ISSUE THREAD

**Issue by wyshi** [facebookresearch/ParlAI#3116]
**Bug description**
1) If `--init-model` contains the state_dict of the optimizer, `torch_generator_agent` will load the optimizer state_dict and ignores the `-lr` set in the command line and set the learning rate to be the initial_lr in the optimizer state_dict, https://github.com/facebookresearch/ParlAI/blob/master/parlai/core/torch_generator_agent.py#L538.

2) Another potential problem is that, in `_init_warmup_scheduler`, https://github.com/facebookresearch/ParlAI/blob/master/parlai/nn/lr_scheduler.py#L55-L64, 
`if self.hard_reset is True and states.get('warmup_scheduler') is not None`, we'd load the warmup_scheduler anyway `self.warmup_scheduler.load_state_dict(states['warmup_scheduler'])`
which could also mess up the warm_up_scheduler and surprise the user?

**Reproduction steps**
use any `--init-model <any model that saves previous optimizer state_dict in the state, the model name shouldn't end with .checkpoint> -lr 2.37e-05` 

**The optimizers looks like**
before loading
```
MemoryEfficientFP16Adam (
Parameter Group 0
    amsgrad: True
    betas: (0.9, 0.999)
    eps: 1e-08
    lr: 2.37e-05
    weight_decay: 0
)
```
after loading optimizer_state_dict
```
MemoryEfficientFP16Adam (
Parameter Group 0
    amsgrad: True
    betas: (0.9, 0.999)
    eps: 1e-08
    initial_lr: 5e-05
    lr: 1.5e-05
    weight_decay: 0
)
```
**Expected behavior**
expected the optimizer and lr_scheduler to start with the learning rate set in the command line.

**Additional context**
torch 1.4.0, parlai 0.9.2

**Comment by shuuki4** [comment]
I'm also experiencing this issue while fine-tuning. In addition to the `optimizer_state_dict`, `warmup_scheduler` state seems to be also being overridden by previous warmup scheduler as well.

**Comment by stephenroller** [comment]
I believe there's a workaround by using pytorch 1.5 or newer

**Comment by shuuki4** [comment]
For my case, I am testing with pytorch 1.6.0 (w/ parlai version 0.9.3) and still experiencing this issue. 😢

**Comment by stephenroller** [comment]
Sad!

**Comment by stephenroller** [comment]
@shuuki4 would resetting the full optimizer work for you?

**Comment by shuuki4** [comment]
@stephenroller Yes, for my case, I just modified a code little bit to skip the optimizer & warmup scheduler state dict loading part in my fine-tuning and it solved above issue. However, it skips the loading of other optimizer parameters as well, so I guess this is not a proper solution :(

**Comment by stephenroller** [comment]
The problem is it's hard to distinguish between the LR scheduler state we want to pull in and that which we don't :(

If you wanted, you could manually edit the state_dict to remove the lr_scheduler and num_training_steps keys, but leave the optimizer and model keys.

**Comment by sjscotti** [comment]
I am having the same problem with learning rate not being what I set in the command line, so I searched and found this issue.  Since the latest  comment is over a year old, the issue doesn't seem resolved.  However, has a satisfactory work around been identified?

**Comment by stephenroller** [comment]
It should be resolved when using --init-model but not --model-file. A workaround is to vacuum the model (parlai vacuum --help) before restarting training.

**Comment by sjscotti** [comment]
Thanks Stephen!  The vacuum did the trick!

**Comment by wise-east** [comment]
Awesome, vacuuming worked for me too! 

I'm curious though to see if I can make the using only --init-model method work. It didn't seem to work when I had both --init-model and --model-file in the command line. 
What's the expected behavior when using --init-model but not --model-file? I thought --model-file would be needed for saving the newly trained model

**Comment by stephenroller** [comment]
The behavior SHOULD be as follows (modulo new bugs):

- try initializing from model_file.checkpoint, the last saved file. If that succeeds, blindly trust everything inside it and assume this is resuming training
- failing that, initialize from model_file. Blindly trust whatever is inside it. Assume we are resuming training, but not saving checkpoints to save time/disk space. This is particularly common for very large files.
- failing that, initialize *only the weights* from init_model. Throw away anything from optimizer state or LR schedulers. Assume we are starting a fresh fine tuning run
- failing that, initialize everything from scratch and assume we're training a fresh model

**Comment by stephenroller** [comment]
I think this is stale enough to disregard
